File: src/gittins_reference/encoding.py
"""Feature encoding: everything is hashed (design doc D2).

The engine never models "arm number 7" — and after this module, it never
models "the price field" either. Context and action features arrive as
plain dicts of whatever the caller has; every feature, the arm's identity,
the intercept, and every interaction term hashes into one fixed space of
2**bits dimensions. The single up-front declaration is `bits`. There is no
schema, no field registration, and nothing to migrate: new field names, new
categorical values, and new arms are all usable the moment they first
appear, and evidence on anything that stops appearing decays out of the
sums, recycling its dimensions (R1). Fixed state memory under unbounded
feature and arm churn, by construction.

**Tokens.** Each supplied (name, value) becomes one token in its namespace
("c|" context, "a|" action):

    string value          ->  token "ns|name=value", contribution 1.0
    numeric value         ->  token "ns|name",       contribution float(value)
    None                  ->  skipped (absent)

so categoricals need no declared value lists — the value is part of the
token. The arm's identity is one more token, "i|arm_id", contribution 1.0:
identity really is just another feature (the original VW move). Bools count
as numeric (True = 1.0).

**The encoding is an outer product of token lists.** With a linear model, a
context-only dimension adds the same amount to every candidate's score and
can never reorder candidates, so context must interact with action features
to matter. Both sides get a leading bias token (the empty string), and
every left x right pair contributes:

    index = hash(left_token, right_token)  masked to bits
    x[index] += sign(hash) * left_value * right_value

The bias tokens make the pieces fall out of the one rule: bias x bias is the
intercept, bias x action the action main effects, context x bias the context
main effects, the rest the context-action interactions. The pair hash is
FNV-1a over the two token strings joined by a separator byte, finished with
the splitmix64 mixer (both from rng.py); the top hash bit gives the sign.

**Collisions are the accepted cost, not an error.** Distinct pairs may land
in the same slot and their contributions add; the sign makes colliding
features as often opposed as conflated, so collisions blur rather than
bias. The cost is bounded by `bits` — size the space generously; linear
state is cheap. Hashes depend only on the tokens, never on a salt or an
agent, so every agent encodes identically and merged sums stay aligned (D3).

**Determinism.** Features are processed in sorted token order, so the
accumulation order — and therefore every output bit — is independent of
dict insertion order, identical on every platform.

The reference materializes the dense 2**bits vector for clarity; the
compiled core will keep the handful of nonzero (index, value) pairs sparse.
"""

from gittins_reference.rng import fnv1a_64, mix64

# Separator between the two tokens of a pair; no printable token contains it.
PAIR_SEP = b"\x1f"


def feature_tokens(namespace: str, values: dict) -> "list[tuple[str, float]]":
    """(token, contribution) per supplied feature, in sorted token order."""
    out = []
    for name in sorted(values):
        v = values[name]
        if v is None:
            continue
        if isinstance(v, str):
            out.append((f"{namespace}|{name}={v}", 1.0))
        elif isinstance(v, (int, float)):
            out.append((f"{namespace}|{name}", float(v)))
        else:
            raise ValueError(f"feature {name!r} has unsupported type {type(v).__name__}")
    return out


def pair_hash(left_token: str, right_token: str) -> int:
    return mix64(fnv1a_64(left_token.encode("utf-8") + PAIR_SEP + right_token.encode("utf-8")))


def encode(context: dict, arm_id: str, action: dict, bits: int) -> list[float]:
    """The feature vector for one (context, candidate) pair: the hashed
    outer product described in the module docstring. The model dimension is
    2**bits (pass to `new_bandit`)."""
    if not (1 <= bits <= 24):
        raise ValueError("bits must be between 1 and 24")
    mask = (1 << bits) - 1
    left = [("", 1.0)] + feature_tokens("c", context)
    right = [("", 1.0)] + feature_tokens("a", action) + [(f"i|{arm_id}", 1.0)]
    x = [0.0] * (1 << bits)
    for left_token, left_value in left:
        for right_token, right_value in right:
            h = pair_hash(left_token, right_token)
            sign = 1.0 if (h >> 63) == 0 else -1.0
            x[h & mask] += sign * left_value * right_value
    return x

File: core/src/encoding.rs
//! Fully hashed feature encoding — the port of `encoding.py`.
//!
//! Features arrive as (name, value) pairs; every feature, categorical
//! value, arm identity, intercept, and interaction hashes into one 2^bits
//! space via the outer product ([bias]+context) x ([bias]+action+identity).
//! Tokens are processed in sorted name order and per-slot contributions
//! accumulate in outer-product iteration order, so every output value is
//! bit-identical to the reference's.

use crate::error::Error;
use crate::rng::{fnv1a_extend, mix64, FNV_START};

/// Separator between the two tokens of a pair; no printable token contains it.
const PAIR_SEP: u8 = 0x1f;

/// A feature value as the public dict-shaped API supplies it. Booleans count
/// as numeric (true = 1.0); `None` means absent (the feature is skipped).
#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    Str(String),
    Num(f64),
    None,
}

/// A sparse candidate: (index, value) pairs in strictly increasing index
/// order, values nonzero — the format every core layer consumes.
pub type Features = Vec<(usize, f64)>;

/// (token, contribution) per supplied feature, in sorted token order.
pub fn feature_tokens(namespace: &str, values: &[(String, Value)]) -> Vec<(String, f64)> {
    let mut sorted: Vec<&(String, Value)> = values.iter().collect();
    sorted.sort_by(|a, b| a.0.cmp(&b.0));
    let mut out = Vec::with_capacity(sorted.len());
    for (name, value) in sorted {
        match value {
            Value::None => {}
            Value::Str(s) => out.push((format!("{namespace}|{name}={s}"), 1.0)),
            Value::Num(v) => out.push((format!("{namespace}|{name}"), *v)),
        }
    }
    out
}

/// The FNV-1a accumulator after folding `left_token` and the separator:
/// the shared prefix state of every pair hash with this left token, folded
/// once per left token instead of once per pair.
fn left_prefix(left_token: &str) -> u64 {
    fnv1a_extend(fnv1a_extend(FNV_START, left_token.as_bytes()), &[PAIR_SEP])
}

/// The 64-bit hash of one (left token, right token) pair: FNV-1a over the
/// tokens joined by the separator byte, finished with the splitmix64 mixer.
/// Streamed through the accumulator — the same byte fold as hashing the
/// joined string, with no intermediate buffer.
pub fn pair_hash(left_token: &str, right_token: &str) -> u64 {
    mix64(fnv1a_extend(left_prefix(left_token), right_token.as_bytes()))
}

/// The left half of the outer product — the bias token plus the context's
/// tokens. The context is candidate-invariant, so callers encoding a whole
/// candidate set compute this once per decision and pass it to
/// `encode_with_context`, instead of rebuilding (sort + format) the same
/// list once per candidate.
pub fn context_tokens(context: &[(String, Value)]) -> Vec<(String, f64)> {
    let mut left = vec![(String::new(), 1.0)];
    left.extend(feature_tokens("c", context));
    left
}

/// The sparse feature vector for one (context, candidate) pair — the hashed
/// outer product, as (index, value) pairs in strictly increasing index
/// order, exact zeros absent. The model dimension is 2^bits.
pub fn encode(
    context: &[(String, Value)],
    arm_id: &str,
    action: &[(String, Value)],
    bits: u32,
) -> Result<Features, Error> {
    encode_with_context(&context_tokens(context), arm_id, action, bits)
}

/// `encode` with the context's token list already built (`context_tokens`):
/// the same outer product over the same tokens in the same order, so the
/// output is bit-identical to `encode`'s — only the per-candidate rebuild
/// of the left half is saved.
pub fn encode_with_context(
    left: &[(String, f64)],
    arm_id: &str,
    action: &[(String, Value)],
    bits: u32,
) -> Result<Features, Error> {
    if !(1..=24).contains(&bits) {
        return Err(Error::new("bits must be between 1 and 24"));
    }
    let mask = (1u64 << bits) - 1;
    // Each right token as the byte parts feature_tokens would have
    // concatenated — bias "", then "a|name" / "a|name=value" in sorted
    // name order, then "i|arm" — folded through the accumulator per pair
    // and never materialized as a string. Same bytes, same order, same
    // hashes; what disappears is one String allocation per action token
    // per candidate per decision.
    const EMPTY: &[u8] = b"";
    let mut sorted: Vec<&(String, Value)> = action.iter().collect();
    sorted.sort_by(|a, b| a.0.cmp(&b.0));
    let mut right: Vec<([&[u8]; 4], f64)> = Vec::with_capacity(sorted.len() + 2);
    right.push(([EMPTY, EMPTY, EMPTY, EMPTY], 1.0)); // the bias token ""
    for (name, value) in sorted {
        match value {
            Value::None => {}
            Value::Str(s) => {
                right.push(([b"a|", name.as_bytes(), b"=", s.as_bytes()], 1.0))
            }
            Value::Num(v) => right.push(([b"a|", name.as_bytes(), EMPTY, EMPTY], *v)),
        }
    }
    right.push(([b"i|", arm_id.as_bytes(), EMPTY, EMPTY], 1.0));
    // The outer product, flat: (slot, contribution) in iteration order,
    // then a stable sort by slot. Stability keeps equal slots in encounter
    // order, so summing each run left to right performs the exact addition
    // sequence the per-slot accumulator did — same bits, no tree of nodes.
    let mut pairs: Vec<(usize, f64)> = Vec::with_capacity(left.len() * right.len());
    for (left_token, left_value) in left {
        let prefix = left_prefix(left_token);
        for (parts, right_value) in &right {
            let mut folded = prefix;
            for part in parts {
                folded = fnv1a_extend(folded, part);
            }
            let h = mix64(folded);
            let sign = if h >> 63 == 0 { 1.0 } else { -1.0 };
            pairs.push(((h & mask) as usize, sign * left_value * right_value));
        }
    }
    pairs.sort_by_key(|&(slot, _)| slot); // stable: encounter order per slot
    let mut out: Features = Vec::with_capacity(pairs.len());
    for (slot, v) in pairs {
        match out.last_mut() {
            Some(last) if last.0 == slot => last.1 += v,
            _ => out.push((slot, v)),
        }
    }
    out.retain(|&(_, v)| v != 0.0);
    Ok(out)
}
